sbibm/algorithms/abcpy/rejection_abc.py

```diff
@@ -0,0 +1,141 @@
+        num_samples: Number of samples to generate from posterior
+        num_simulations: Simulation budget
+        num_simulations_per_param: The number of the total simulations budget to produce for each parameter value;
+            used to better estimate the ABC distance from the observation for each parameter value. Useful to
+            investigate whether splitting the budget to do multi-simulations per parameter is helpful or not.
+        num_observation: Observation number to load, alternative to `observation`
+        observation: Observation, alternative to `num_observation`
+        num_top_samples: If given, will use `top=True` with num_top_samples
+        quantile: Quantile to use
+        eps: Epsilon threshold to use
+        distance: Distance to use; can be "l2" (Euclidean Distance), "log_reg" (LogReg) or "pen_log_reg" (PenLogReg)
+        save_distances: If True, stores distances of samples to disk
+        sass: If True, summary statistics are learned as in
+            Fearnhead & Prangle 2012.
+        sass_fraction: Fraction of simulation budget to use for sass.
+        sass_feature_expansion_degree: Degree of polynomial expansion of the summary
+            statistics.
+        kde_bandwidth: If not None, will resample using KDE when necessary, set
+            e.g. to "cv" for cross-validated bandwidth selection
+    Returns:
+        Samples from posterior, number of simulator calls, log probability of true params if computable
+    """
+    if sass:
+        raise NotImplementedError("SASS not yet implemented")
+
+    assert not (num_observation is None and observation is None)
+    assert not (num_observation is not None and observation is not None)
+
+    if (eps is not None) + (num_top_samples is not None) + (quantile is not None) != 1:
+        raise RuntimeError("Exactly one of `num_top_samples`, `quantile`, `eps` needs to be specified.")
+
+    log = sbibm.get_logger(__name__)
+    log.info(f"Running REJ-ABC")
+
+    if observation is None:
+        observation = task.get_observation(num_observation)
+
+    if num_top_samples is not None:
+        if sass:
+            quantile = num_top_samples / (
+                    num_simulations - int(sass_fraction * num_simulations)
+            )
+        else:
+            quantile = num_top_samples / num_simulations
+
+    # use the dummy backend only (for now)
+    backend = BackendDummy()
+
+    # as Rejection ABC works by directly drawing until you get below some threshold -> here we set the threshold to
+    # very large value and then we select the best objects later.
+
+    statistics = Identity(degree=1, cross=True)  # we assume statistics are already computed inside the task
+    distance_calc = get_distance(distance, statistics)
+
+    # define prior and model from the task simulator:
+    parameters = ABCpyPrior(task)
+    model = ABCpySimulator([parameters], task, max_calls=num_simulations)
+
+    # inference; not sure how to use random seed
```

Comment:
Not sure how seeding is handled in `abcpy` but perhaps it's a good idea to introduce an argument that allows passing it?

Reply:
Basically when you instantiate an inference algorithm in ABCpy (which is a class) you pass to it a random seed which is used to initialize a random number generator which is then used everywhere around the inference class. 

I have added a seed option to the `run` function which is used for that. I only did now know whether adding that or not as the other algorithms you implemented do not have it. 

I believe even if you use the abcpy `run` function without explicitly passing the seed but you instead beforehand fix the numpy random seed, reproducibility should be preserved, but the seed can be useful anyway.

Reply:
Actually using the seed in ABCpy does not ensure reproducibility without doing at the start of the script
```
    torch.manual_seed(1)
    np.random.seed(1)
```
that is due to the fact that the simulators do not take a random seed or random number generator as inputs. On the other hand, fixing the above without the ABCpy seed ensures reproducibility. 

Therefore I removed the `seed` parameter from the `run` function for ABCpy, as it was useless. That has therefore the same API as all others.